scripts/install-windows-prereqs.ps1

```diff
@@ -516,23 +529,38 @@ function Install-DCAP-Dependencies {
     Install-Tool -InstallerPath $PACKAGES["dcap"]["local_file"] `
                  -ArgumentList @('/auto', "$PACKAGES_DIRECTORY\Intel_SGX_DCAP")
 
+    $OS_VERSION = Get-WindowsRelease
     if (($LaunchConfiguration -eq "SGX1FLC") -or ($LaunchConfiguration -eq "SGX1FLC-NoDriver") -or ($DCAPClientType -eq "Azure"))
     {
         $drivers = @{
-            'sgx_base_dev' = @{
-                'zip_path'    = "$PACKAGES_DIRECTORY\Intel_SGX_DCAP\Intel SGX DCAP for Windows *\LC_driver_WinServer2016\Signed_*.zip"
-                'location'    = 'root\SgxLCDevice'
-                'description' = 'Intel(R) Software Guard Extensions Launch Configuration Service'
+            'WinServer2016' = @{
+                'sgx_base_dev' = @{
+                    'zip_path'    = "$PACKAGES_DIRECTORY\Intel_SGX_DCAP\Intel SGX DCAP for Windows *\LC_driver_${OS_VERSION}\Signed_*.zip"
+                    'location'    = 'root\SgxLCDevice'
+                    'description' = 'Intel(R) Software Guard Extensions Launch Configuration Service'
+                }
+                'sgx_dcap_dev' = @{
+                    'zip_path'    = "$PACKAGES_DIRECTORY\Intel_SGX_DCAP\Intel SGX DCAP for Windows *\DCAP_INF\${OS_VERSION}\Signed_*.zip"
+                    'location'    = 'root\SgxLCDevice_DCAP'
+                    'description' = 'Intel(R) Software Guard Extensions DCAP Components Device'
+                }
             }
-            'sgx_dcap_dev' = @{
-                'zip_path'    = "$PACKAGES_DIRECTORY\Intel_SGX_DCAP\Intel SGX DCAP for Windows *\DCAP_INF\WinServer2016\Signed_*.zip"
-                'location'    = 'root\SgxLCDevice_DCAP'
-                'description' = 'Intel(R) Software Guard Extensions DCAP Components Device'
+            'WinServer2019' = @{
+                'sgx_base' = @{
+                    'zip_path'    = "$PACKAGES_DIRECTORY\Intel_SGX_DCAP\Intel SGX DCAP for Windows *\LC_driver_${OS_VERSION}\Signed_*.zip"
+                    'location'    = 'root\SgxLCDevice'
```

Comment:
This is an ACPI device not a SW device on 16299 and later, should be 

acpi\int0e0c

Reply:
@fchinchilla Could you elaborate on the line that should be changed?

Reply:
Line 551, instead of 'root\...'
it should be 'acpi\int0e0c'

Reply:
Hello @fchinchilla!

We tried your suggested change (https://github.com/oprinmarius/openenclave/commit/b48e1db967d26dd4b2653c457d18afd61740608b) but all the tests on Windows Server 2019 fail with:
```
...
2 - tests/safecrt (Exit code 0xc0000139)
3 - tests/safecrt-lvi-cfg (Exit code 0xc0000139)
8 - tests/switchless_ocalls (Exit code 0xc0000139)
9 - tests/switchless_ocalls-lvi-cfg (Exit code 0xc0000139)
...
```

NOTE: These only pop up for the DCAP Windows Server 2019 nodes. The non-DCAP nodes are fine with the device rename.

Reverting the change, and using the same device name as on Windows Server 2016 didn't cause any failures with the DCAP testing stages, and everything is fine.

Reply:
Not sure about the tests, but when I try to replicate the installation portion of the script on a 2019 VM without internet access, the results look like this:

PS C:\Users\Administrator\Downloads> .\devcon.exe install '.\Intel SGX DCAP for Windows v1.5.100.2\LC_driver_WinServer2019\Signed_1152921504628335119\drivers\da362676-240a-4ec6-98cf-4cc4430c84be\sgx_base.inf' 'root\SGXLCDevice'
Device node created. Install is complete when drivers are installed...
Updating drivers for root\SGXLCDevice from C:\Users\Administrator\Downloads\Intel SGX DCAP for Windows v1.5.100.2\LC_driver_WinServer2019\Signed_1152921504628335119\drivers\da362676-240a-4ec6-98cf-4cc4430c84be\sgx_base.inf.
devcon.exe failed.

PS C:\Users\Administrator\Downloads> $LASTEXITCODE
2

PS C:\Users\Administrator\Downloads> .\devcon.exe install '.\Intel SGX DCAP for Windows v1.5.100.2\LC_driver_WinServer2019\Signed_1152921504628335119\drivers\da362676-240a-4ec6-98cf-4cc4430c84be\sgx_base.inf' 'acpi\int0e0c'
Device node created. Install is complete when drivers are installed...
Updating drivers for acpi\int0e0c from C:\Users\Administrator\Downloads\Intel SGX DCAP for Windows v1.5.100.2\LC_driver_WinServer2019\Signed_1152921504628335119\drivers\da362676-240a-4ec6-98cf-4cc4430c84be\sgx_base.inf.
Drivers installed successfully.

I would be surprised the install worked, since the .inf files have:

[SGXPSW_LC.NTamd64.10.0.3..16299]
%SGXPSW_LC.DeviceDesc% = SGXPSW_LC_Device, acpi\int0e0c 

and

[SgxDCAPComponentsDev.NTamd64.10.0]
%SgxDCAPComponentsnDevice.DeviceDesc% = SgxDCAPComponentsDevice, SWC\VEN_INT&DEV_0E0C_DCAP

and devcon AFAIK cannot install a driver meant for device 'foo' to be used for device 'bar', the ID needs to match.

I'll see if I can take a closer look at the tests.

Reply:
We check the `$LASTEXITCODE` for the devcon install:
https://github.com/oprinmarius/openenclave/blob/2f5f51b25c62c081287410a3950626635e736385/scripts/install-windows-prereqs.ps1#L595-L598

However, when we generate the CI Windows Server 2019 image, we have internet access and we run Windows Update to setup the Intel SGX driver.

In our PowerShell script, the `devcon.exe` install part is conditioned by `$LaunchConfiguration` parameter being `SGX1FLC` (which is only set for Windows Server 2016):
https://github.com/oprinmarius/openenclave/blob/2f5f51b25c62c081287410a3950626635e736385/scripts/install-windows-prereqs.ps1#L592-L600

And with WS 2019, the `$LaunchConfiguration` is `SGX1FLC-NoDriver`, so the `devcon.exe` install part is skipped, because we don't want to overwrite anything already installed by Windows updates.

But, I double checked the script to figure out why we had failures when we added your suggested changed. It looks like, before the conditioned devcon install part, we remove the driver:
https://github.com/oprinmarius/openenclave/blob/2f5f51b25c62c081287410a3950626635e736385/scripts/install-windows-prereqs.ps1#L581-L591

We used to do this, because on Windows Server 2016 we didn't rely on Windows Update for the driver setup (which as far as I understood is not the case on Windows Server 2019), and we wanted to make sure existing driver is uninstalled before installing a new driver.

So, on Windows Server 2019:
* with the proper driver IDs (as suggested by you), the driver gets removed, presumably already installed by Windows Update since tests work without any problem, and it remains removed, and we end up with failures.
* without the proper drivers IDs, the driver is not found installed, and the removal part is skipped, and we end up with a working image with the driver setup from Windows Update.

To fix this inconsistency with Windows Server 2019, the removal part should definitely go inside the conditioned `devcon.exe` install command.
 
Thus, the device IDs won't be used anywhere with Windows Server 2019 (because we rely on Windows Update).

On WS 2019, we only copy some DLLs to `$($env:SystemRoot)\system32`:
https://github.com/oprinmarius/openenclave/blob/2f5f51b25c62c081287410a3950626635e736385/scripts/install-windows-prereqs.ps1#L601-L605

Reply:
Thanks for explaining it. I saw the check for $LASTEXITCODE but missed the fact that 2019 was not supporting SGX1FLC.